File: src/Python/12.IntegerToRoman.py

class Solution:
    def intToRoman(self, num):
        """
        :type num: int
        :rtype: str
        """
        roman = 'MDCLXVI'
        romandiv = [1000, 500, 100, 50, 10, 5, 1]
        ans = ''   
        strnum = str(num)
        i = 6 - (len(strnum) - 1) * 2
        for c in strnum:
            if c == '1':
                ans += roman[i]
            elif c == '2':
                ans += roman[i] * 2
            elif c == '3':
                ans += roman[i] * 3
            elif c == '4':
                ans += roman[i] + roman[i - 1]
            elif c == '5':
                ans += roman[i - 1]
            elif c == '6':
                ans += roman[i - 1] + roman[i]
            elif c == '7':
                ans += roman[i - 1] + roman[i] * 2
            elif c == '8':
                ans += roman[i - 1] + roman[i] * 3
            elif c == '9':
                ans += roman[i] + roman[i - 2]
            i += 2
        return ans


if __name__ == "__main__":
    solution = Solution()
    print(solution.intToRoman(3))
    print(solution.intToRoman(4))
    print(solution.intToRoman(9))
    print(solution.intToRoman(58))
    print(solution.intToRoman(1994))
   

File: src/Java/12.IntegerToRoman.java

import java.awt.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.Queue;
import java.util.HashSet;


class Solution {
    public String intToRoman(int num) {
        String res = "";
        int[] tokenkey = new int[]{1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
        String[] tokenval = new String[]{"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};
        for(int i = 0;i < 13;++i){
            int n = tokenkey[i];
            String t = tokenval[i];
            while (num >= n){
                res += t;
                num -= n;
            }
        }
        return res;
    }
}
